haskell_scotty/Server.hs

```diff
@@ -1,24 +1,15 @@
 {-# LANGUAGE OverloadedStrings #-}
-import Web.Scotty
-import Network.HTTP.Types.Status
-
--- | Build a basic test page.
-helloTestPage :: ActionM ()
-helloTestPage = html "Hello world"
-
--- | Build a test page and do some status stuff.
-compoundTestPage :: ActionM ()
-compoundTestPage = do
-  html "You are not supposed to be here."
-  status badRequest400
 
--- | Specify the behavior of each API endpoint.
-router :: ScottyM ()
-router = do
-  get "/api/test" helloTestPage
-  get "/api/test2" compoundTestPage
+import Web.Scotty
+import FetchClients
+import Control.Monad.IO.Class
 
--- | Construct the actual webserver.
 main :: IO ()
-main = scotty 6969 router
+main = scotty 6969 $ do
+  get "/api/test" $ do
+    html "Hello world"
+
+  get "/api/clients" $ do
+    clients <- liftIO fetchClients
```

Comment:
If you look at the version I keep writing :-) where those inner `do` blocks are split out into separate functions with their own types, it will become more obvious. The `do` in question is actually in an `ActionM` monad. Since `ActionM` is built on top of `IO`, you can use IO functions there, but you have to "strip off" the rest of the `ActionM` stuff for `fetchClients` explicitly. It's basically a type conversion thing.